### DISCUSSION TITLE
Brainstorm to orgazine plugins and rules

### DISCUSSION THREAD

**Discussion by maykon-oliveira** [riccardoperra/lexical-angular#36] [General]
I was taking a look at lexical implementation for react, and many plugins are not distribuited together. So, for example, if I'm thinking to add a toolbar with some plugins in my own project, I need copy the source code from `lexical-playground` library and paste it in my code.

I dont know if they still architecting the lib and yours dependencies, but it not make sense for me.

Many plugins like ToolbarPlugin, TreeViewPlugin, TwitterPlugin, YouTubePlugin, ImagesPlugin, EmojisPlugin... should be in `@lexical-react`. What do u think?

The *HorizontalRulePlugin* for example, I puted it in lexical-angular, but following the lexial repository structure, it should be in playground.

I was thinking to update docs too, a guide explaing how to add a new plugin in angular. For example:

Every new custom plugin should have:
- A Directive:
    - [Command. ](https://github.com/riccardoperra/lexical-angular/pull/33/files#diff-2eb102ac1023de752e57ba235eb9d760dbf974d85b22d081ecdae3319cdb29f8R16)[Lexical Reference.](https://lexical.dev/docs/concepts/commands#createcommand)
    - [Check if node is enabled in editor](https://github.com/riccardoperra/lexical-angular/pull/33/files#diff-2eb102ac1023de752e57ba235eb9d760dbf974d85b22d081ecdae3319cdb29f8R30)
    - [Command Listener. ](https://github.com/riccardoperra/lexical-angular/pull/33/files#diff-2eb102ac1023de752e57ba235eb9d760dbf974d85b22d081ecdae3319cdb29f8R37)[Lexical Reference](https://lexical.dev/docs/concepts/commands#editorregistercommand)
- A Rule:
    - [Decorator Node. ](https://github.com/riccardoperra/lexical-angular/pull/33/files#diff-b6ef311489c5746ef17135f97ec2cb96a6d949e56af0549419a3fef78f7822a4R19)[Lexical Reference.](https://lexical.dev/docs/concepts/nodes#decoratornode)
- [Component](https://github.com/riccardoperra/lexical-angular/pull/33/files#diff-556694b1efcbd82809867e9222431632b56d603c157320f0e63556368cda1468R53)

**Comment by riccardoperra** [comment]
Hey, I think this issue can be moved to `Discussion`. What do you think? 




> I was taking a look at lexical implementation for react, and many plugins are not distribuited together. So, for example, if I'm thinking to add a toolbar with some plugins in my own project, I need copy the source code from `lexical-playground` library and paste it in my code.
> 
> I dont know if they still architecting the lib and yours dependencies, but it not make sense for me.

They may have changed their workflow, but when I started to work to this adapter, their intention was to separate the react implementation to the "core" which is headless.

At the moment they don't have on npm the @lexical-playground package so yes, each time you have to copy and paste the source code.

> Many plugins like ToolbarPlugin, TreeViewPlugin, TwitterPlugin, YouTubePlugin, ImagesPlugin, EmojisPlugin... should be in @lexical-react. What do u think?

I agree. But I think this is what they want because I believe that for them this is a library from which to start to build your editor.

> The HorizontalRulePlugin for example, I puted it in lexical-angular, but following the lexial repository structure, it should be in playground.

If we follow their repository yes, but it doesn't make much sense to me right now. Now I'd just port what's on @lexical/react, while ready-made components might be an alternative thing (toolbar, etc.). Nothing prevents us from doing things differently

In `lexical-angular` I called the folder `plugins` actually, but in fact the porting for now has been done only of the components of @lexical/react right?

**Comment by maykon-oliveira** [comment]
Yes, make sense. Let's move to Discussion.

For while, we can just port what's on @lexical/react. There are some open issues to add more features, I'll move forward it.

**Comment by riccardoperra** [comment]
I made a checklist of components/plugins already ported and to port 😄 I believe there is everything  https://github.com/riccardoperra/lexical-angular/issues/37
